internal/client/client_interceptor.go

```diff
@@ -34,10 +36,54 @@ func NewInterceptedClient(client Client) Client {
 }
 
 func (r interceptedClient) SendRequest(ctx context.Context, addr string, req *tikvrpc.Request, timeout time.Duration) (*tikvrpc.Response, error) {
+	// Build the resource control interceptor if there is one and the resource group name is given.
+	var rcInterceptor interceptor.RPCInterceptor
+	resourceGroupName := req.GetResourceGroupName()
+	if resourceControlInterceptor != nil && len(resourceGroupName) > 0 {
+		rcInterceptor = func(next interceptor.RPCInterceptorFunc) interceptor.RPCInterceptorFunc {
+			return func(target string, req *tikvrpc.Request) (*tikvrpc.Response, error) {
+				reqInfo := resource_control.MakeRequestInfo(req)
+				err := resourceControlInterceptor.OnRequestWait(ctx, resourceGroupName, reqInfo)
+				if err != nil {
+					return nil, err
+				}
+				resp, err := next(target, req)
+				if resp != nil {
+					respInfo := resource_control.MakeResponseInfo(resp)
+					resourceControlInterceptor.OnResponse(ctx, resourceGroupName, reqInfo, respInfo)
+				}
+				return resp, err
+			}
+		}
+	}
+	// Chain the interceptors if there are multiple interceptors.
+	var finalInterceptor interceptor.RPCInterceptor
+	if rcInterceptor != nil {
+		finalInterceptor = rcInterceptor
+	}
 	if it := interceptor.GetRPCInterceptorFromCtx(ctx); it != nil {
-		return it(func(target string, req *tikvrpc.Request) (*tikvrpc.Response, error) {
+		if finalInterceptor != nil {
+			finalInterceptor = interceptor.ChainRPCInterceptors(finalInterceptor, it)
+		} else {
+			finalInterceptor = it
+		}
+	}
+	if finalInterceptor != nil {
+		return finalInterceptor(func(target string, req *tikvrpc.Request) (*tikvrpc.Response, error) {
 			return r.Client.SendRequest(ctx, target, req, timeout)
 		})(addr, req)
 	}
 	return r.Client.SendRequest(ctx, addr, req, timeout)
 }
+
+var resourceControlInterceptor client.ResourceGroupKVInterceptor
```

Comment:
should use atomic? I want to dynamically control the interceptor behavior associate with the `tidb_enable_resource_control` global variable.